stl/inc/ranges

```diff
@@ -6171,6 +6171,26 @@ namespace ranges {
         inline constexpr _Slide_fn slide;
     } // namespace views
 
+    template <class _Pr>
+    struct _Negated_pred {
+        template <class _Ty1, class _Ty2>
+        _NODISCARD constexpr bool operator()(_Ty1&& _Left, _Ty2&& _Right) {

```

Comment:
Given https://github.com/microsoft/STL/issues/2889#issuecomment-1191736156, this should instead be

```suggestion
        _NODISCARD constexpr decltype(auto) operator()(_Ty1&& _Left, _Ty2&& _Right) {
```

Reply:
As @hewillk mentions in #2889, the result is converted to `bool` in `adjacent_find` anyway, so returning `bool` should have the same end result as returning `decltype(auto)`.

An argument against `decltype(auto)` is: if the return type is `decltype(auto)`, the function body will need to be instantiated for `invoke_result_t` to get the actual return type, which is extra work.

(I also considered `decltype(!_STD invoke(_Pred, _STD forward<_Ty1>(_Left), _STD forward<_Ty2>(_Right)))`, but that's ugly and unnecessary.)

Reply:
Returning `delctype(auto)` is verbose and unnecessary. Returning `bool` better reflects the characteristics of the "predicate".

Reply:
mmh, okay. Then this needs to `static_cast<bool>(...)` in order to avoid explicit and implicit conversions resulting in different values;

```cxx
template <class Pr>
constexpr bool if_(Pr pr) {
    if (pr()) {
        return true;
    } else {
        return false;
    }
}

template <class Pr>
constexpr bool impl(Pr pr) {
    return pr();
}

template <class Pr>
constexpr bool check_if_impl(Pr pr) {
    return if_(pr) == impl(pr);
}

struct Bool {
    constexpr explicit operator bool() {
        return true;
    }
    constexpr operator int() {
        return 0;
    }
};

int main() {
    static_assert(check_if_impl([] { return Bool{}; })); // results in a failure
}
```

Reply:
A `Bool` with only `explicit operator bool` does not model _`boolean-testable`_, which means we shouldn't do an explicit conversion here.

Reply:
This _is_ implicitly convertible to `bool`, it just goes through a conversion to `int` first.